[dcp] Add more granularity on what collectives to apply for rank-local checkpointing

ankitageorge opened the pull request:
Summary: Currently we only have a boolean to represent whether we do collectives on checkpointing. But we should have more granularity on this, to be able to skip only certain collectives. This is specifically useful to be able to skip the planning collective on anchor checkpoints, but continue to do the metadata collective so that we can output the anchor checkpoint in the format that can be used by downstream users. The planning collective is the more expensive of the two, so we will still see some of the wins for anchor checkpoints.

Test Plan: ensure existing tests pass

Differential Revision: D86347143




cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci

Skylion007 commented on torch/distributed/checkpoint/_enums.py:
```diff
@@ -0,0 +1,14 @@
+# mypy: allow-untyped-defs
```
Why not just type this file, it's just an enum. Should already be done?
```suggestion
```

ankitageorge committed:
[dcp] Add more granularity on what collectives to apply for rank-local checkpointing (#167234)

Summary:

Currently we only have a boolean to represent whether we do collectives on checkpointing. But we should have more granularity on this, to be able to skip only certain collectives. This is specifically useful to be able to skip the planning collective on anchor checkpoints, but continue to do the metadata collective so that we can output the anchor checkpoint in the format that can be used by downstream users. The planning collective is the more expensive of the two, so we will still see some of the wins for anchor checkpoints.

Test Plan: ensure existing tests pass

Differential Revision: D86347143
